### PR TITLE
Multilisten in opensource

### PR DESCRIPTION
**Pull request by EvgenyMekhanik**
iproto: implement ability to open several listening sockets
Previously, iproto could only open one socket for listening
only. This patch change this behaviour, now user can open
several listening sockets (up to 20). Also in addition to
ability to pass uri as a number or string, as  previously,
ability to pass uri as a table of numbers or strings has
been added.
```lua
box.cfg { listen = {3301, 3302, 3303} }
box.cfg { listen = {"127.0.0.1:3301", "127.0.0.1:3302"} }
```

### CODE REVIEW

File: src/box/iproto.cc [hunk 1]
```diff
@@ -257,15 +255,16 @@ unsigned iproto_readahead = 16320;
...
+char **
+iproto_bound_address(int *size)
 {
-	if (tx_binary.addr_len == 0)
-		return NULL;
-	sio_addr_snprintf(buf, SERVICE_NAME_MAXLEN,
-			  (struct sockaddr *)&tx_binary.addrstorage,
-			  tx_binary.addr_len);
-	return buf;
+	return evio_service_bound_address(&tx_binary, size);
```
Comment (Gerold103) [inline src/box/iproto.cc:261]: This is quite a complicated way to return address array you have chosen. Please, drop these mallocs/frees/copying, try this (you will need to add comments and fix potential bugs since I didn't run tests on this):
```Diff
diff --git a/src/box/iproto.cc b/src/box/iproto.cc
index eab6ce825..21869a4f0 100644
--- a/src/box/iproto.cc
+++ b/src/box/iproto.cc
@@ -255,16 +255,19 @@ unsigned iproto_readahead = 16320;
 /* The maximal number of iproto messages in fly. */
 static int iproto_msg_max = IPROTO_MSG_MAX_MIN;
 
-char **
-iproto_bound_address(int *size)
+int
+iproto_addr_count(void)
 {
-	return evio_service_bound_address(&tx_binary, size);
+	return evio_service_count(&tx_binary);
 }
 
-void
-iproto_bound_address_free(char **bound_adress_array)
+const char *
+iproto_addr_str(char *buf, int idx)
 {
-	return evio_service_bound_address_free(bound_adress_array);
+	socklen_t size;
+	const struct sockaddr *addr = evio_service_addr(&tx_binary, idx, &size);
+	sio_addr_snprintf(buf, SERVICE_NAME_MAXLEN, addr, size);
+	return buf;
 }
 
 /**
diff --git a/src/box/iproto.h b/src/box/iproto.h
index 99af39882..2afb67495 100644
--- a/src/box/iproto.h
+++ b/src/box/iproto.h
@@ -93,22 +93,11 @@ iproto_thread_stats_get(struct iproto_stats *stats, int thread_id);
 void
 iproto_reset_stat(void);
 
-/**
- * Allocate and return array, which contains representation of the
- * addresses served by iproto. To be shown in box.info. Buffer should
- * be freed by the caller using 'iproto_bound_address_free'. @a size
- * contains count of served addresses. Each address contained in block
- * with SERVICE_NAME_MAXLEN len.
- */
-char **
-iproto_bound_address(int *size);
+int
+iproto_addr_count(void);
 
-/**
- * Free @a bound_address_array allocated by
- * 'iproto_bound_address' function.
- */
-void
-iproto_bound_address_free(char **bound_address_array);
+const char *
+iproto_addr_str(char *buf, int idx);
 
 int
 iproto_rmean_foreach(void *cb, void *cb_ctx);
diff --git a/src/box/lua/info.c b/src/box/lua/info.c
index 17e9603d0..b46e1bebd 100644
--- a/src/box/lua/info.c
+++ b/src/box/lua/info.c
@@ -584,20 +584,21 @@ lbox_info_sql(struct lua_State *L)
 static int
 lbox_info_listen(struct lua_State *L)
 {
-	int size;
-	char **bound_address_array = iproto_bound_address(&size);
-	if (size == 0) {
+	int count = iproto_addr_count();
+	if (count == 0) {
 		lua_pushstring(L, NULL);
-	} else if (size == 1) {
-		lua_pushstring(L, bound_address_array[0]);
-	} else {
-		lua_createtable(L, size, 0);
-		for (int i = 0; i < size; i++) {
-			lua_pushstring(L, bound_address_array[i]);
-			lua_rawseti(L, -2, i + 1);
-		}
+		return 1;
+	}
+	char addrbuf[SERVICE_NAME_MAXLEN];
+	if (count == 1) {
+		lua_pushstring(L, iproto_addr_str(addrbuf, 0));
+		return 1;
+	}
+	lua_createtable(L, count, 0);
+	for (int i = 0; i < count; i++) {
+		lua_pushstring(L, iproto_addr_str(addrbuf, i));
+		lua_rawseti(L, -2, i + 1);
 	}
-	iproto_bound_address_free(bound_address_array);
 	return 1;
 }
 
diff --git a/src/lib/core/evio.c b/src/lib/core/evio.c
index b2d097d93..505a082ae 100644
--- a/src/lib/core/evio.c
+++ b/src/lib/core/evio.c
@@ -63,16 +63,6 @@ evio_service_entry_is_active(struct evio_service_entry *entry)
 	return entry->ev.fd >= 0;
 }
 
-static inline int
-evio_service_entry_bound_address(const struct evio_service_entry *entry,
-				 char *buf)
-{
-	const struct sockaddr *sockaddr = (struct sockaddr *)
-		&entry->addrstorage;
-	return sio_addr_snprintf(buf, SERVICE_NAME_MAXLEN, sockaddr,
-				 entry->addr_len);
-}
-
 /** Note: this function does not throw. */
 void
 evio_close(ev_loop *loop, struct ev_io *evio)
@@ -467,30 +457,19 @@ evio_service_reuse_addr(const char **uris, int size)
 	return 0;
 }
 
-char **
-evio_service_bound_address(const struct evio_service *service, int *size)
+int
+evio_service_count(const struct evio_service *service)
 {
-	*size = service->entry_count;
-	if (service->entry_count == 0)
-		return NULL;
-	char **bound_address_array = (char **)
-		xcalloc(service->entry_count,
-			sizeof(char *) + SERVICE_NAME_MAXLEN);
-	char *dstbuf = (char *)bound_address_array +
-		service->entry_count * sizeof(char *);
-	for (int i = 0; i < service->entry_count; i++) {
-		bound_address_array[i] =
-			dstbuf + i * SERVICE_NAME_MAXLEN;
-		evio_service_entry_bound_address(&service->entries[i],
-						 bound_address_array[i]);
-	}
-	return bound_address_array;
+	return service->entry_count;
 }
 
-void
-evio_service_bound_address_free(char **bound_address_array)
+const struct sockaddr *
+evio_service_addr(const struct evio_service *service, int idx, socklen_t *size)
 {
-	free(bound_address_array);
+	assert(idx < service->entry_count);
+	const struct evio_service_entry *e = &service->entries[idx];
+	*size = e->addr_len;
+	return &e->addr;
 }
 
 void
diff --git a/src/lib/core/evio.h b/src/lib/core/evio.h
index 00f612546..6b98664bf 100644
--- a/src/lib/core/evio.h
+++ b/src/lib/core/evio.h
@@ -85,19 +85,11 @@ struct evio_service {
         ev_loop *loop;
 };
 
-/**
- * Return array of strings which represented addresses served by @a service.
- * Save array size in @a size.
- */
-char **
-evio_service_bound_address(const struct evio_service *service, int *size);
+int
+evio_service_count(const struct evio_service *service);
 
-/**
- * Free @a bound_address_array allocated by
- * 'evio_service_bound_address' function.
- */
-void
-evio_service_bound_address_free(char **bound_address_array);
+const struct sockaddr *
+evio_service_addr(const struct evio_service *service, int idx, socklen_t *size);
 
 /** Initialize the service. Don't bind to the port yet. */
 void
```

File: src/box/iproto.cc [hunk 2]
```diff
@@ -185,9 +185,7 @@ struct iproto_thread {
 	 * Iproto thread id
 	 */
 	uint32_t id;
-	/*
-	 * Iproto binary listener
-	 */
+	/** Array of iproto binary listeners */
```
Comment (Gerold103) [inline src/box/iproto.cc:188]: Please, extract `evio_service` turn into an array into a separate commit. Otherwise you are mixing refactoring (`evio_service`) with a public feature (multilisten). It is hard to tell which changes in this commit belong to what.

File: changelogs/unreleased/gh-3554-several-listening-sockets.md [hunk 3]
```diff
@@ -0,0 +1,5 @@
+## feature/core
+* Implemented ability to open several listening sockets.
```
Comment (Gerold103) [inline changelogs/unreleased/gh-3554-several-listening-sockets.md:2]: Could you please add an empty line before this line?

File: src/box/iproto.cc [hunk 4]
```diff
@@ -2501,7 +2500,7 @@ iproto_on_accept(struct evio_service *service, int fd,
 	struct iproto_msg *msg;
 
 	struct iproto_thread *iproto_thread =
-		(struct iproto_thread*)service->on_accept_param;
+		(struct iproto_thread *)service->on_accept_param;
```
Comment (Gerold103) [inline src/box/iproto.cc:2503]: This does not seem to belong to this commit.

File: src/box/iproto.cc [hunk 5]
```diff
@@ -2788,15 +2787,9 @@ iproto_init(int threads_count)
...
-
+	evio_service_init(loop(), &tx_binary, 0, "tx_binary", NULL, NULL);
 	iproto_threads = (struct iproto_thread *)
-		calloc(threads_count, sizeof(struct iproto_thread));
-	if (iproto_threads == NULL) {
-		tnt_raise(OutOfMemory, threads_count *
-			  sizeof(struct iproto_thread), "calloc",
-			  "struct iproto_thread");
-	}
+		xcalloc(threads_count, sizeof(struct iproto_thread));
```
Comment (Gerold103) [inline src/box/iproto.cc:2792]: Ditto. It is not related to this commit.

File: src/lib/core/coio.cc [hunk 6]
```diff
@@ -680,7 +680,8 @@ coio_service_init(struct coio_service *service, const char *name,
 void
 coio_service_start(struct evio_service *service, const char *uri)
 {
-	if (evio_service_bind(service, uri) != 0 ||
+	const char *uris[] = { uri };
+	if (evio_service_bind(service, uris) != 0 ||
```
Comment (Gerold103) [inline src/lib/core/coio.cc:684]: That is a strange API. So you need to pass count in `init()` but then always pass the same number of uris in `bind`? Maybe better make `bind` take a uri array with its size? I don't see any pros in passing count to `init()`. But I do see that it is quite easy to get problems if you pass an array of wrong size in `bind()` forgetting to update the `init()` call.

File: src/lib/core/evio.c [hunk 7]
```diff
@@ -39,6 +39,40 @@
...
+}
+
+static inline int
+evio_service_entry_bound_address(const struct evio_service_entry *entry,
+				 char *buf)
+{
+	const struct sockaddr *sockaddr = (struct sockaddr *)
+		&entry->addrstorage;
+	return sio_addr_snprintf(buf, SERVICE_NAME_MAXLEN, sockaddr,
+				 entry->addr_len);
```
Comment (Gerold103) [inline src/lib/core/evio.c:73]:
```Diff
@@ -67,9 +67,7 @@ static inline int
 evio_service_entry_bound_address(const struct evio_service_entry *entry,
 				 char *buf)
 {
-	const struct sockaddr *sockaddr = (struct sockaddr *)
-		&entry->addrstorage;
-	return sio_addr_snprintf(buf, SERVICE_NAME_MAXLEN, sockaddr,
+	return sio_addr_snprintf(buf, SERVICE_NAME_MAXLEN, &entry->addr,
 				 entry->addr_len);
```

File: src/lib/core/evio.c [hunk 8]
```diff
@@ -202,38 +238,41 @@ evio_service_accept_cb(ev_loop *loop, ev_io *watcher, int events)
...
 		return -1;
 	}
-	int save_errno = errno;
-	int cl_fd = sio_socket(service->addr.sa_family, SOCK_STREAM, 0);
+	if (strncmp(u.host, URI_HOST_UNIX, u.host_len) != 0)
+		return 0;
+
+	struct sockaddr_un un;
+	snprintf(un.sun_path, sizeof(un.sun_path), "%.*s",
+		 (int) u.service_len, u.service);
```
Comment (Gerold103) [inline src/lib/core/evio.c:257]:
```Diff
 	snprintf(un.sun_path, sizeof(un.sun_path), "%.*s",
-		 (int) u.service_len, u.service);
+		 (int)u.service_len, u.service);
```

File: src/lib/core/evio.c [hunk 9]
```diff
@@ -202,38 +238,41 @@ evio_service_accept_cb(ev_loop *loop, ev_io *watcher, int events)
...
+	int cl_fd = sio_socket(un.sun_family, SOCK_STREAM, 0);
 	if (cl_fd < 0)
 		return -1;
 
-	if (connect(cl_fd, &service->addr, service->addr_len) == 0)
+	if (connect(cl_fd, (struct sockaddr *)&un, sizeof(un)) == 0)
 		goto err;
 
-	if (errno != ECONNREFUSED)
+	if (errno == ECONNREFUSED  && unlink(un.sun_path) != 0)
```
Comment (Gerold103) [inline src/lib/core/evio.c:267]: Double whitespace before `&&`.

File: src/lib/core/evio.c [hunk 10]
```diff
@@ -294,69 +324,64 @@ evio_service_bind_addr(struct evio_service *service)
  *
  * @retval 0 for success
  */
-int
-evio_service_listen(struct evio_service *service)
+static int
+evio_service_entry_listen(struct evio_service_entry *entry)
```
Comment (Gerold103) [inline src/lib/core/evio.c:328]: I propose to call it `..._create()`. `init` is a wrong suffix which we have a chance to fix here alongside in all places since these lines are changed anyway.

File: src/lib/core/evio.c [hunk 11]
```diff
@@ -367,59 +392,178 @@ evio_service_bind(struct evio_service *service, const char *uri)
...
 		say_error("Failed to unlink unix "
 			  "socket path: %s", strerror(errno));
 	}
 }
 
+static void
+evio_service_entry_attach(struct evio_service_entry *dst,
+			 const struct evio_service_entry *src)
+{
+	strcpy(dst->host, src->host);
```
Comment (Gerold103) [inline src/lib/core/evio.c:454]: Please, add an assertion that this entry does not have a socket already.

File: src/lib/core/evio.c [hunk 12]
```diff
@@ -367,59 +392,178 @@ evio_service_bind(struct evio_service *service, const char *uri)
...
+	dst->addr_len = src->addr_len;
+	ev_io_set(&dst->ev, src->ev.fd, EV_READ);
+}
+
+static inline int
+evio_service_reuse_addr(const char **uris, int size)
+{
+	for (int i = 0; i < size; i++)
+		if (evio_service_entry_reuse_addr(uris[i]) != 0)
+			return -1;
```
Comment (Gerold103) [inline src/lib/core/evio.c:466]: For multiline `for` loops you need to wrap its body into `{}`.

File: src/lib/core/evio.c [hunk 13]
```diff
@@ -367,59 +392,178 @@ evio_service_bind(struct evio_service *service, const char *uri)
...
+	for (int i = 0; i < service->entry_count; i++)
+		evio_service_entry_init(&service->entries[i], service);
+}
+
+void
+evio_service_attach(struct evio_service *dst, const struct evio_service *src)
+{
+	for (int i = 0; i < src->entry_count; i++)
+		evio_service_entry_attach(&dst->entries[i], &src->entries[i]);
+	dst->entry_count = src->entry_count;
```
Comment (Gerold103) [inline src/lib/core/evio.c:517]: So does it mean the counts could be not the same? Then how does the loop above work when it is not the same? It will go out of array bounds and crash, won't it? Or some of evio entries will just be left out of the borders and not freed ever in case the count is decreased.

File: src/lib/core/evio.c [hunk 14]
```diff
@@ -367,59 +392,178 @@ evio_service_bind(struct evio_service *service, const char *uri)
...
+	service->entry_count = 0;
+	service->entries = NULL;
+}
+
+bool
+evio_service_is_active(struct evio_service *service)
+{
+	for (int i = 0; i < service->entry_count; i++)
+		if (evio_service_entry_is_active(&service->entries[i]))
+			return true;
```
Comment (Gerold103) [inline src/lib/core/evio.c:535]: You would need `{}` for the loop body when it has > 1 line. Also `service` can be made `const` here. Together with entry argument in `evio_service_entry_is_active`.

File: src/lib/core/evio.c [hunk 15]
```diff
@@ -367,59 +392,178 @@ evio_service_bind(struct evio_service *service, const char *uri)
...
+	return false;
+}
+
+int
+evio_service_listen(struct evio_service *service)
+{
+	for (int i = 0; i < service->entry_count; i++)
+		if (evio_service_entry_listen(&service->entries[i]) != 0)
+			return -1;
```
Comment (Gerold103) [inline src/lib/core/evio.c:544]: `{}` for the loop body.

File: src/lib/core/sio.c [hunk 16]
```diff
@@ -215,7 +215,7 @@ int
 sio_bind(int fd, struct sockaddr *addr, socklen_t addrlen)
 {
 	int rc = bind(fd, addr, addrlen);
-	if (rc < 0 && errno != EADDRINUSE)
```
Comment (Gerold103) [inline src/lib/core/sio.c:218]: Does not seem related to this commit. Why isn't it done in the previous one?

File: test/box/several-listening-sockets.result [hunk 17]
```diff
@@ -0,0 +1,173 @@
+-- test-run result file version 2
```
Comment (Gerold103) [inline test/box/several-listening-sockets.result:1]: Isn't it obligatory to use either TAP or luatest now instead of diff tests?

File: test/box/several-listening-sockets.result [hunk 18]
```diff
@@ -0,0 +1,173 @@
...
+-- according to template @a addr. It's simple adds
+-- capital letters in alphabetical order to the
+-- template.
+function create_uri_table(addr, count)
+    local uris_table = {}
+    local path_table = {}
+    local ascii_A = 97
+    for i = 1, count do
+        local ascii_code = ascii_A + i - 1
+        local letter = string.upper(string.char(ascii_code))
```
Comment (Gerold103) [inline test/box/several-listening-sockets.result:29]: You don't need `string.upper`:
```Diff
-    local ascii_A = 97
+    local ascii_A = string.byte('A')
     for i = 1, count do
         local ascii_code = ascii_A + i - 1
-        local letter = string.upper(string.char(ascii_code))
+        local letter = string.char(ascii_code)
         path_table[i] = addr .. letter
```

File: test/box/several-listening-sockets.result [hunk 19]
```diff
@@ -0,0 +1,173 @@
...
+        path_table[i] = addr .. letter
+        uris_table[i] = "unix/:" .. addr .. letter
+    end
+    return uris_table, path_table
+end;
+ | ---
+ | ...
+function check_connection(port)
```
Comment (Gerold103) [inline test/box/several-listening-sockets.result:37]: Please, leave empty lines between functions.

File: test/box/several-listening-sockets.result [hunk 20]
```diff
@@ -0,0 +1,173 @@
...
+ | ...
+box.cfg{listen = port_number}
+ | ---
+ | ...
+check_connection(box.info.listen)
+ | ---
+ | ...
+
+-- Check connection if listening uri passed as a single string.
+unix_socket_path = "./tarantool" .. "A"
```
Comment (Gerold103) [inline test/box/several-listening-sockets.result:61]: Too much formatting and concatenations, it could be much simpler.
```Diff
 -- Check connection if listening uri passed as a single string.
-unix_socket_path = "./tarantool" .. "A"
-box.cfg{listen = "unix/:" .. unix_socket_path}
-assert(box.cfg.listen == "unix/:" .. unix_socket_path)
+unix_socket_path = "unix/:./tarantoolA"
+box.cfg{listen = unix_socket_path}
+assert(box.cfg.listen == unix_socket_path)
 assert(box.info.listen:match("unix/:"))
-check_connection("unix/:" .. unix_socket_path)
+check_connection(unix_socket_path)
```

File: test/box/several-listening-sockets.result [hunk 21]
```diff
@@ -0,0 +1,173 @@
...
+-- so we need to make sure that all resources for those uris for which
+-- everything has already completed will be successfully cleared in case
+-- of error for one of the next uri in list.
+uri_table, path_table = create_uri_table("./tarantool", 5)
+ | ---
+ | ...
+uri_table[#uri_table + 1] = "baduri:1"
+ | ---
+ | ...
+uri_table[#uri_table + 1] = "unix/:" .. "./tarantool" .. "X"
```
Comment (Gerold103) [inline test/box/several-listening-sockets.result:133]:
```Diff
@@ -69,8 +69,8 @@ assert(fio.path.exists(default_listen_addr))
 -- everything has already completed will be successfully cleared in case
 -- of error for one of the next uri in list.
 uri_table, path_table = create_uri_table("./tarantool", 5)
-uri_table[#uri_table + 1] = "baduri:1"
-uri_table[#uri_table + 1] = "unix/:" .. "./tarantool" .. "X"
+table.insert(uri_table, "baduri:1")
+table.insert(uri_table, "unix/:./tarantoolX")
 
 -- can't resolve uri for bind
 box.cfg{listen = uri_table}
@@ -83,7 +83,7 @@ test_run:cmd("push filter 'error: bind, called on fd [0-9]+' \
               to 'error: bind, called on fd <number>'")
 
 -- Special test case when we try to listen several identical URIs
-uri = "unix/:" .. "./tarantool"
+uri = "unix/:./tarantool"
 box.cfg{listen = {uri, uri, uri}}
```

File: test/box/several-listening-sockets.result [hunk 22]
```diff
@@ -0,0 +1,173 @@
...
+end
+ | ---
+ | ...
+assert(box.cfg.listen == default_listen_addr)
+ | ---
+ | - true
+ | ...
+
+test_run:cmd("push filter 'error: bind, called on fd [0-9]+' \
+              to 'error: bind, called on fd <number>'")
```
Comment (Gerold103) [inline test/box/several-listening-sockets.result:153]: You don't really need the filter:
```Diff
@@ -79,11 +79,9 @@ for i, path in ipairs(path_table) do \
 end
 assert(box.cfg.listen == default_listen_addr)
 
-test_run:cmd("push filter 'error: bind, called on fd [0-9]+' \
-              to 'error: bind, called on fd <number>'")
-
 -- Special test case when we try to listen several identical URIs
 uri = "unix/:" .. "./tarantool"
-box.cfg{listen = {uri, uri, uri}}
+ok, err = pcall(box.cfg, {listen = {uri, uri, uri}})
+assert(not ok and err.message:match('error: bind'))
 assert(not fio.path.exists(uri))
```

**Review by Gerold103 — CHANGES_REQUESTED** [review]
Thanks for the patchset!

File: src/lib/core/evio.c [hunk 23]
```diff
@@ -301,18 +301,8 @@ evio_service_listen(struct evio_service *service)
 		  sio_strfaddr(&service->addr, service->addr_len));
```
[the remark this answers is not in the document]
Comment (EvgenyMekhanik) [reply src/lib/core/evio.c]: fixed

File: changelogs/unreleased/gh-6535-fix-error-in-listening-to-numerical-uri-after-listening-to-unix-socket.md [hunk 24]
```diff
@@ -0,0 +1,4 @@
+## bugfix/core
+
+* Fix error in listening when user pass uri in numerical form
```
[the remark this answers is not in the document]
Comment (EvgenyMekhanik) [reply changelogs/unreleased/gh-6535-fix-error-in-listening-to-numerical-uri-after-listening-to-unix-socket.md:3]: fixed

File: test/box-tap/several-listening-sockets.test.lua [hunk 25]
```diff
@@ -0,0 +1,93 @@
...
+end
+local function check_connection(port)
+    local conn = net_box.connect(port)
+    local rc = conn:ping()
+    conn:close()
+    return rc
+end
+
+local test = tap.test('gh-6535-listen-update-numeric-uri')
+test:plan(1 + 3 + 5 + (4 * 5 + 5) + 7 + 3)
```
Comment (Gerold103) [inline test/box-tap/several-listening-sockets.test.lua:31]: Maybe just put a single number here? This notation does not simplify its update really. Anyway update nearly always happens by running the test, looking at the number in the failure message, and putting it into the plan.

File: test/box-tap/several-listening-sockets.test.lua [hunk 26]
```diff
@@ -0,0 +1,93 @@
...
+test:ok(not box.cfg.listen, "box.cfg.listen")
+
+-- Special test case when we try to listen several identical URIs
+local uri = "unix/:./tarantool"
+local ok, err = pcall(box.cfg, {listen = {uri, uri, uri}})
+test:ok(not ok and err.message:match('bind'), "err.message:match")
+test:ok(not fio.path.exists(uri), "fio.path.exists")
+test:ok(not box.cfg.listen, "box.cfg.listen")
+
+os.exit(test:check() and 0 or 1)
```
Comment (Gerold103) [inline test/box-tap/several-listening-sockets.test.lua:93]: Please, add an empty line in the end of the file.

File: test/box-tap/several-listening-sockets.test.lua [hunk 27]
```diff
@@ -0,0 +1,93 @@
...
+-- so we need to make sure that all resources for those uris for which
+-- everything has already completed will be successfully cleared in case
+-- of error for one of the next uri in list.
+local uri_table, path_table = create_uri_table("./tarantool", 5)
+table.insert(uri_table, "baduri:1")
+table.insert(uri_table, "unix/:./tarantoolX")
+
+-- can't resolve uri for bind
+local ok, err = pcall(box.cfg, {listen = uri_table})
+test:ok(not ok and err.message:match('can\'t resolve uri for bind'),
```
Comment (Gerold103) [inline test/box-tap/several-listening-sockets.test.lua:79]: You can use `""` and drop the escaping `\`.

**Review by Gerold103 — CHANGES_REQUESTED** [review]
Thanks for the fixes!

**Review by Gerold103 — APPROVED** [review]
